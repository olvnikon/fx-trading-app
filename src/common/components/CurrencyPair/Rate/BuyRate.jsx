// @flow
import React from 'react';
import Rate from './Rate';
import styles from './Rate.scss';

type Props = {|
  +currency: string,
  +rate: number,
|};

const BuyRate = ({ currency, ...rest }: Props) => (
  <div className={styles.buyRate}>
    <Rate header={`Buy ${currency}`} {...rest} />
  </div>
);

export default BuyRate;
