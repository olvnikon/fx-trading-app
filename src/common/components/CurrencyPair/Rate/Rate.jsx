// @flow
import React from 'react';
import styles from './Rate.scss';

type Props = {|
  +rate: number,
  +header: string,
|};

const mapRightSide = chunks => (
  chunks.map((chunk, index) => (
    <span
      key={index} // eslint-disable-line react/no-array-index-key
      className={styles[`chunk-${index + 1}`]}
    >
      {chunk}
    </span>
  )));

// ToDo: refactor from imperative to declarative
const mapRate = (rate) => {
  const [leftSide, rightSide] = rate.toString().split('.');
  const leftLayout = <span>{`${leftSide}.`}</span>;
  const chunks = rightSide.match(/.{1,2}/g) || ['00'];
  const rightLayout = mapRightSide(chunks);
  return <span>{leftLayout}{rightLayout}</span>;
};

const Rate = ({ rate, header }: Props) => (
  <section className={styles.rate}>
    <header>
      {header}
    </header>
    <main>
      {mapRate(rate)}
    </main>
  </section>
);

export default Rate;
