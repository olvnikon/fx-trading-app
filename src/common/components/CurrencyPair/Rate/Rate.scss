@import "../../../styles/variables";
$rateBackground: #fff;
$sellRateBackground: #fc0017;
$sellRateColor: #c70721;
$buyRateBackground: #82d146;
$buyRateColor: #698934;
$rateHeight: 50px;
$rateBorderWidth: 2px;
$rateArrowBorderWidth: $rateHeight / 2;
$rateWidth: 80%;
$rateActiveBorder: #4fadc5;
$borderRadius: 5px;

.rate {
  background-color: $rateBackground;
  border: $rateBorderWidth solid $currencyBorder;
  box-sizing: border-box;
  height: $rateHeight;
  padding: 0.3em;
  width: $rateWidth;

  header {
    font-size: 0.5em;
    font-weight: 700;
  }

  main {
    font-size: 1em;
  }

  .chunk-2 {
    font-size: 1.5em;
  }

  .chunk-3 {
    vertical-align: top;
  }
}

.sellRate {
  color: $sellRateColor;
  float: left;
  position: relative;
  text-align: left;

  .rate {
    border-radius: $borderRadius 0 0 $borderRadius;
    border-right-width: 0;
    float: left;
  }

  &:after, &:before {
    left: $rateWidth;
    top: 50%;
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &:after {
    border-left-color: $rateBackground;
    border-width: $rateArrowBorderWidth - $rateBorderWidth;
    margin-top: -($rateArrowBorderWidth - $rateBorderWidth);
  }

  &:before {
    border-left-color: $currencyBorder;
    border-width: $rateArrowBorderWidth;
    margin-top: -$rateArrowBorderWidth;
  }

  &:hover {
    .rate {
      background: $sellRateBackground;
      border: $rateBorderWidth solid $rateActiveBorder;
      border-right-width: 0;
      color: $rateBackground;
    }

    &:after {
      border-left-color: $sellRateBackground;
    }

    &:before {
      border-left-color: $rateActiveBorder;
    }
  }
}

.buyRate {
  color: $buyRateColor;
  right: 0;
  position: relative;
  text-align: right;

  .rate {
    border-radius: 0 $borderRadius $borderRadius 0;
    border-left-width: 0;
    float: right;
  }

  &:after, &:before {
    right: $rateWidth;
    top: 50%;
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &:after {
    border-right-color: $rateBackground;
    border-width: $rateArrowBorderWidth - $rateBorderWidth;
    margin-top: -($rateArrowBorderWidth - $rateBorderWidth);
  }

  &:before {
    border-right-color: $currencyBorder;
    border-width: $rateArrowBorderWidth;
    margin-top: -$rateArrowBorderWidth;
  }

  &:hover {
    .rate {
      background: $buyRateBackground;
      border: $rateBorderWidth solid $rateActiveBorder;
      border-left-width: 0;
      color: $rateBackground;
    }

    &:after {
      border-right-color: $buyRateBackground;
    }

    &:before {
      border-right-color: $rateActiveBorder;
    }
  }
}
