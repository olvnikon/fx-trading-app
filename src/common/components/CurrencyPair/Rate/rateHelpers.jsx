// @flow
import React from 'react';
import styles from './Rate.scss';

const mapRightSide = chunks => (
  chunks.map((chunk, index) => (
    <span
      key={index} // eslint-disable-line react/no-array-index-key
      className={styles[`chunk-${index + 1}`]}
    >
      {chunk}
    </span>
  )));

// ToDo: refactor from imperative to declarative
// eslint-disable-next-line import/prefer-default-export
export const mapRate = (rate: number) => {
  const [leftSide, rightSide] = rate.toString().split('.');
  const leftLayout = <span>{`${leftSide}.`}</span>;
  const chunks = rightSide.match(/.{1,2}/g) || ['00'];
  const rightLayout = mapRightSide(chunks);
  return <span>{leftLayout}{rightLayout}</span>;
};
