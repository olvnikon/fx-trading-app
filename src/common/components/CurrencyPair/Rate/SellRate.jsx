// @flow
import React from 'react';
import Rate from './Rate';
import styles from './Rate.scss';

type Props = {|
  +currency: string,
  +rate: number,
|};

const SellRate = ({ currency, ...rest }: Props) => (
  <div className={styles.sellRate}>
    <Rate header={`Sell ${currency}`} {...rest} />
  </div>
);

export default SellRate;
