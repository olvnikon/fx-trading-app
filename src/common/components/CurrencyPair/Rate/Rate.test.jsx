// @flow
import React from 'react';
import { shallow } from 'enzyme';
import Rate from './Rate';

describe('Rate', () => {
  const RATE = 12.345;
  const HEADER = 'USD';

  test('should match snapshot', () => {
    const props = {
      rate: RATE,
      header: HEADER,
    };
    const component = shallow(<Rate {...props} />);
    expect(component).toMatchSnapshot();
  });
});
