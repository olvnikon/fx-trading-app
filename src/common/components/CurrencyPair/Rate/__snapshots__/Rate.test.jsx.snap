// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate should match snapshot 1`] = `
<section>
  <header>
    USD
  </header>
  <main>
    <span>
      <span>
        12.
      </span>
      <span
        key="0"
      >
        34
      </span>
      <span
        key="1"
      >
        5
      </span>
    </span>
  </main>
</section>
`;
